src/main/java/com/minecolonies/coremod/commands/ActionArgumentType.java

```diff
@@ -0,0 +1,391 @@
+        for (final Colony colony : colonyList)
+        {
+            final List<CitizenData> citizenDataList = colony.getCitizenManager().getCitizens();
+            for (final CitizenData citizenData : citizenDataList)
+            {
+                citizenNameList.add(citizenData.getName());
+            }
+        }
+        return citizenNameList;
+    }
+
+    public List<String> getTabCompletions(@NotNull final MinecraftServer server,
+            @Nullable final BlockPos pos,
+            final String potentialArgumentValue)
+    {
+        switch (this)
+        {
+            case INTEGER:
+                return Collections.emptyList();
+            case BOOLEAN:
+                return Arrays.asList(new String[] {"true", "false"});
+            case COORDINATE_X:
+            case COORDINATE_Y:
+            case COORDINATE_Z:
+                return getCoordinateTabCompletions(pos, potentialArgumentValue);
+            case PLAYER:
+                final List<String> playerNameStrings = getOnlinePlayerNames(server);
+                return playerNameStrings.stream().filter(k -> k.startsWith(potentialArgumentValue)).collect(Collectors.toList());
+            case COLONY:
+                return getColonyTabCompletions(potentialArgumentValue);
+            case CITIZEN:
+                return getCitizenTabCompletions(potentialArgumentValue);
+            default:
+                throw new IllegalStateException("Unimplemented ActionArgumentType tab completion");
+        }
+    }
+
+    private List<String> getCoordinateTabCompletions(@Nullable final BlockPos pos, final String potentialArgumentValue)
+    {
+        if (null == pos)
+        {
+            return Collections.emptyList();
+        }
+        if (potentialArgumentValue.isEmpty())
+        {
+            switch (this)
+            {
+                case COORDINATE_X:
+                    return Collections.singletonList(String.valueOf(pos.getX()));
+                case COORDINATE_Y:
+                    return Collections.singletonList(String.valueOf(pos.getY()));
+                case COORDINATE_Z:
+                    return Collections.singletonList(String.valueOf(pos.getZ()));
+                default:
+                    // We will never reach here.
+                    break;
+            }
+            return Collections.emptyList();
+        }
+        try
+        {
+            Integer.parseInt(potentialArgumentValue);
+            return Collections.emptyList();
+        }
+        catch (final NumberFormatException e)
+        {
+            return Collections.emptyList();
+        }
+    }
+
+    private List<String> getColonyTabCompletions(final String potentialArgumentValue)
+    {
+        final List<String> colonyNumberStrings = getColonyIdStrings();
+        if (potentialArgumentValue.isEmpty())
+        {
+            return colonyNumberStrings;
+        }
+        try
+        {
+            Integer.parseInt(potentialArgumentValue);
```

Comment:
The try-parseint-catch-block is very PERFORMANCE HEAVY. Please change to using a regex.

Reply:
This parseInt doesn't run every tick.  It only runs when a user executes a command or hits tab.  I doubt it has a measurable performance impact in this context.

Reply:
It does not matter. A tr-parseint-catch is about a million times more performace heavy then a simple regex match against \d+. Everytime the JVM has the enter a try it has to save the entire stack and create a readonly-write-on-commit copy of the memory the try block might touch. When a catch is entered this is unwinded which takes even more performance.

Please change this to use a Regular Expression instead. It is simply a very bad practice.

Reply:
@OrionDevelopment 

Disregarding that you haven't used regular expressions to convert ints anywhere else in the code base, nor fixed the existing parseInt() calls.....

Premature optimization certainly matters.   Using faster code that's harder to write and understand comes at a cost as well.   And addressing one kind of performance issue (try-catch) without considering the alternate method's performance issue (allocating objects) is short-sighted.   Using a regular expression creates objects which have to then be garbage-collected.

Here's the results of using parseInt() and pattern.match() exactly one million times:
The good value parsed is "5".  The bad value parsed is "bad".  This is with the Pattern pre-compiled and stored in a static final variable.

First, only returning a boolean to indicate if it's a good or bad value:

parseBy_parseIntYesNo(valueToParseGood) time=18
parseBy_patternMatcherPartialYesNo(valueToParseGood) time=87
parseBy_parseIntYesNo(valueToParseBad) time=1058
parseBy_patternMatcherPartialYesNo(valueToParseBad) time=68

So for good values, parseInt is around 4x faster.
For bad values, pattern matching is around 15x faster.

But unfortunately, just knowing the pattern is good or bad doesn't get you the int.  So we have to call parseInt() even if the pattern matching is good.   And we have to convert it to an Integer to use null to indicate a bad result:

parseInt(valueToParseGood) time=20
patternMatcherPartial(valueToParseGood) time=185
parseInt(valueToParseBad) time=974
patternMatcherPartial(valueToParseBad) time=114

So for good values, parseInt is 6x faster.
For bad values. pattern matching is 8x faster.

So what's the actual "cost" for the worse-case of using parseInt to return a boolean with a value according to these measurements?

1000 / 1,000,000 ms or 1/1000th of a ms or 1 nanosecond.
 
Note that we can't calculate the true worse-case cost of pattern matching as we don't have a way to easily measure how the objects created affect garbage collection.

Note well that the cost of creating and running the try-parseInt-catch block is LESS than the cost of pattern matching.

And then there's the problem of defining an appropriate pattern.   "[-]?\\d+" was the one I used, but it doesn't guarantee that the number will fit into an int.

On top of all that, there are other options than parseInt or pattern matching. 

com.google.common.primitives.Ints:
tryParse(valueToParseGood) time=73
tryParse(valueToParseBad) time=19

tryParse is 6x faster than pattern matching and 55x faster than parseInt when parsing bad values.
tryParse is 2x faster than pattern matching and 3.5x SLOWER than parseInt when parsing good values.
tryParse is about the same as parseInt for detecting a good value and about the same as pattern matching for detecting a bad value.

That makes it a far better choice while maintaining easy readability and insures that the validation used is the same for both parsing ints and matching an int pattern.

And here's one final item to consider, to put things into perspective, the time to "new Object()" a million times:

createObject(valueToParseBad) time=8

So creating the most basic object is only 132x faster than handling an exception in a try-catch block.

```
    private static final String INT_PATTERN_STRING = "[-]?\\d+";
    private static final Pattern INT_PATTERN = Pattern.compile(INT_PATTERN_STRING);

    public Integer parseBy_parseInt(final String valueToParse)
    {
        try
        {
            final int intValue = Integer.parseInt(valueToParse);
            return Integer.valueOf(intValue);
        }
        catch (final NumberFormatException e)
        {
            return null;
        }
    }

    public Integer parseBy_patternMatcherPartial(final String valueToParse)
    {
        final Matcher matcher = INT_PATTERN.matcher(valueToParse);
        final boolean didMatch = matcher.matches();
        if (didMatch)
        {
            final int intValue = Integer.parseInt(valueToParse);
            return Integer.valueOf(intValue);
        }
        else
        {
            return null;
        }
    }


    public Integer parseBy_tryParse(final String valueToParse)
    {
        return Ints.tryParse(valueToParse);
    }

    public boolean parseBy_parseIntYesNo(final String valueToParse)
    {
        try
        {
            Integer.parseInt(valueToParse);
            return true;
        }
        catch (final NumberFormatException e)
        {
            return false;
        }
    }

    public boolean parseBy_patternMatcherPartialYesNo(final String valueToParse)
    {
        final Matcher matcher = INT_PATTERN.matcher(valueToParse);
        return matcher.matches();
    }

    public boolean createObject(final String valueToParse)
    {
        new Object();
        return true;
    }
```

Reply:
Okey I see. I am sorry I was told that a regex was at least faster then a try-catch.
I totally agree that tryParse is the better solution.